agent/api/ecsclient/client.go

```diff
@@ -290,19 +290,75 @@ func (client *APIECSClient) SubmitTaskStateChange(change api.TaskStateChange) er
 	}
 
 	status := change.Status.String()
-	_, err := client.submitStateChangeClient.SubmitTaskStateChange(&ecs.SubmitTaskStateChangeInput{
-		Cluster: &client.config.Cluster,
-		Task:    &change.TaskArn,
-		Status:  &status,
-		Reason:  &change.Reason,
-	})
+
+	req := ecs.SubmitTaskStateChangeInput{
+		Cluster: aws.String(client.config.Cluster),
+		Task:    aws.String(change.TaskArn),
+		Status:  aws.String(status),
+		Reason:  aws.String(change.Reason),
+	}
+
+	containerEvents := make([]*ecs.ContainerStateChange, len(change.Containers))
+	for i, containerEvent := range change.Containers {
+		containerEvents[i] = client.buildContainerStateChangePayload(containerEvent)
+	}
+
+	req.Containers = containerEvents
+
+	_, err := client.submitStateChangeClient.SubmitTaskStateChange(&req)
 	if err != nil {
 		log.Warn("Could not submit a task state change", "err", err)
 		return err
 	}
+
 	return nil
 }
 
+func (client *APIECSClient) buildContainerStateChangePayload(change api.ContainerStateChange) *ecs.ContainerStateChange {
+	statechange := &ecs.ContainerStateChange{
+		ContainerName: aws.String(change.ContainerName),
+	}
+
+	if change.Reason != "" {
+		if len(change.Reason) > ecsMaxReasonLength {
+			trimmed := change.Reason[0:ecsMaxReasonLength]
+			statechange.Reason = aws.String(trimmed)
+		} else {
+			statechange.Reason = aws.String(change.Reason)
+		}
+	}
+	status := change.Status
+
+	if status != api.ContainerStopped && status != api.ContainerRunning {
+		seelog.Warn("Not submitting unsupported upstream container state", "state", status)
```

Comment:
I prefer the formatted string kind because it gives you control over the output format and not having to deal with whatever golang thinks the stringified version of an object should look like.